### PR TITLE
PMP address range & NAPOT decoding bugs

### PR DESCRIPTION
**Pull request by dayeol**
Suggested fix for the issue #149

### CODE REVIEW

File: target/riscv/pmp.c [hunk 1]
```diff
@@ -245,7 +245,7 @@ bool pmp_hart_has_privs(CPURISCVState *env, target_ulong addr,
          from low to high */
     for (i = 0; i < MAX_RISCV_PMPS; i++) {
         s = pmp_is_in_range(env, i, addr);
-        e = pmp_is_in_range(env, i, addr + size);
+        e = pmp_is_in_range(env, i, addr + size - 1);
```
Comment (alistair23) [inline target/riscv/pmp.c]: This doesn't look right. This fixes the less then end address but breaks the greater then or equal to start address.

Shouldn't the diff be this instead?
diff --git a/target/riscv/pmp.c b/target/riscv/pmp.c
index f432f3b759..9eae74715b 100644
--- a/target/riscv/pmp.c
+++ b/target/riscv/pmp.c
@@ -210,7 +210,7 @@ static int pmp_is_in_range(CPURISCVState *env, int pmp_index, target_ulong addr)
     int result = 0;
 
     if ((addr >= env->pmp_state.addr[pmp_index].sa)
-        && (addr <= env->pmp_state.addr[pmp_index].ea)) {
+        && (addr < env->pmp_state.addr[pmp_index].ea)) {
         result = 1;
     } else {
         result = 0;

File: target/riscv/pmp.c [hunk 2]
```diff
@@ -138,7 +138,7 @@ static void pmp_decode_napot(target_ulong a, target_ulong *sa, target_ulong *ea)
         return;
     } else {
         target_ulong t1 = ctz64(~a);
-        target_ulong base = (a & ~(((target_ulong)1 << t1) - 1)) << 3;
+        target_ulong base = (a & ~(((target_ulong)1 << t1) - 1)) << 2;
```
Comment (alistair23) [inline target/riscv/pmp.c]: This looks good. It probably should be split into a separate commit.

File: target/riscv/pmp.c (hunk 1 above)
Comment (dayeol) [reply target/riscv/pmp.c]: Thank you for the review!

However, that will make even more legitimate accesses fail, and also will make it return 0 for some in-PMP-range accesses.
Because `env->pmp_state.addr[pmp_index].ea` is set to the last address, not a boundary address,
even a `pmp_is_in_range` call with the last address itself will always fail.

Let me give you an example, assuming your diff has been applied.
If PMP range is from 0x0010 to 0x0020, `sa` will set to 0x0010 and `ea` will set to 0x001f
Accessing 1 byte at 0x001f will call
`pmp_is_in_range(0x001f)` and 
`pmp_is_in_range(0x0020)`
Since both 0x001f and 0x0020 are not less than `ea` (=0x001f), both call will return 0,
meaning that the address is not in the PMP range, which is not true.

I think `pmp_is_in_range` is doing great, but the issue here is that we are using a wrong address.
Can you explain more about why does the change break the greater then or equal to start address rule?

File: target/riscv/pmp.c (hunk 2 above)
Comment (dayeol) [reply target/riscv/pmp.c]: I'll split the commit into two. Thank you!

File: target/riscv/pmp.c (hunk 1 above)
Comment (alistair23) [reply target/riscv/pmp.c]: Looking at page 47 on the privileged spec I see " the entry matches any address a such that pmpaddri-1 <= a < pmpaddri" which is why I think the pmp_is_in_range() should be changed to less then.

Although I do see your point. It looks like ea already has one subtracted from it when it is set so you are right about it being the boundary.

Comment (alistair23) [reply target/riscv/pmp.c]: Should that 1 not be subtracted when setting ea instead? Then we can have the same comparison that the spec has

**Comment by michaeljclark** [comment]
Just to note the PMP code has not been thoroughly tested. I would like to write some PMP test cases at some point. riscv-tests currently does not have any PMP tests.

Also note there were some recent changes with respect to PMP granularity in the latest draft spec. Currently PMPs are checked in get_physical_address on the TLB slow path so the granularity in QEMU’s PMP implementation is 12-bits (PAGE_SIZE). Will have to find the wording to see how we are supposed to correctly handle this granularity.

File: target/riscv/pmp.c (hunk 1 above)
Comment (dayeol) [reply target/riscv/pmp.c]: I agree. Not subtracting 1 from `ea` should also work.
But note that the spec saying pmpaddri-1 <= a < pmpaddri will be satisfied with both implementations. 
pmpaddri is a riscv register value whereas `ea` is a qemu's variable to configure a pmp range end address.

**Comment by dayeol** [comment]
@michaeljclark Thank you for your comment.
I see. Does that mean we postpone this fix until we have the tests?

Speaking of granularity, can we add the page offset with the pa before we call `pmp_hart_has_privs`?
Thank you.

**Comment by michaeljclark** [comment]
> On 17/07/2018, at 4:52 PM, Dayeol Lee <notifications@github.com> wrote:
> 
> @michaeljclark Thank you for your comment.
> I see. Does that mean we postpone this fix until we have the tests?
> 
> Speaking of granularity, can we add the page offset with the pa before we call pmp_hart_has_privs?
> 

It is static. It will always be TARGET_PAGE_SIZE which is 4096 bytes.

Here is the text from the spec that I was looking for: https://github.com/riscv/riscv-isa-manual/blob/master/src/machine.tex

“\begin{commentary} Software may determine the PMP granularity by writing zero to {\tt pmp0cfg}, then writing all ones to {\tt pmpaddr0}, then reading back {\tt pmpaddr0}. If $G$ is the index of the least-significant bit set, the PMP granularity is $2^{G+2}$ bytes. \end{commentary}”

Given we have to indicate granularity detection via this method, then the PMP code also needs to be statically aware of the granule size. This is a new requirement recently added to the spec so it has not been implemented in the PMP code.

I’m happy to apply fixes to this code but it would be nice to have some test cases... :-D

File: target/riscv/pmp.c (hunk 1 above)
Comment (dayeol) [reply target/riscv/pmp.c]: I found that not substracting 1 from `ea` will complicate the original code
especially when we want to include the entire memory inside a range.
Currently we're assigning -1 (all ones) no matter how large the actual physical memory is,
but if we add 1, then we should know the actual memory size.
Better way around is just pass the end address to `pmp_is_in_range`, 
which is `base + size - 1`

Comment (alistair23) [reply target/riscv/pmp.c]: Ok, in which case all of this looks good to me.

You can add my:
Reviewed-by: Alistair Francis <alistair.francis@wdc.com>
Just copy it below your SOB line. It would be good if you could still split the two patches

Comment (dayeol) [reply target/riscv/pmp.c]: Hi @alistair23,
Should I just PR this again to this repo (riscv-qemu), or report it to upstream QEMU?

Comment (alistair23) [reply target/riscv/pmp.c]: I think you should do both. It makes sense to have it merged here (then it's in and people can use it). It also will eventually need to go upstream so why not do it now :)

QEMU is in freeze though, so it won't be merged straight away.
